File: mobile/src/main/java/org/openhab/habdroid/model/NfcTag.java
package org.openhab.habdroid.model;

import android.net.Uri;
import android.text.TextUtils;
import androidx.annotation.Nullable;

import com.google.auto.value.AutoValue;

@AutoValue
public abstract class NfcTag {
    public static final String SCHEME = "openhab";
    public static final String QUERY_PARAMETER_ITEM_NAME = "i";
    public static final String DEPRECATED_QUERY_PARAMETER_ITEM_NAME = "item";
    public static final String QUERY_PARAMETER_STATE = "s";
    public static final String DEPRECATED_QUERY_PARAMETER_STATE = "command";
    public static final String QUERY_PARAMETER_MAPPED_STATE = "m";
    public static final String QUERY_PARAMETER_ITEM_LABEL = "l";

    @Nullable
    public abstract String sitemap();
    @Nullable
    public abstract String item();
    @Nullable
    public abstract String label();
    @Nullable
    public abstract String state();
    @Nullable
    public abstract String mappedState();

    @AutoValue.Builder
    abstract static class Builder {
        abstract Builder sitemap(String sitemap);
        abstract Builder item(String item);
        abstract Builder label(String label);
        abstract Builder state(String state);
        abstract Builder mappedState(String mappedState);

        abstract NfcTag build();
    }

    public static @Nullable NfcTag fromTagData(Uri uri) {
        if (uri == null || !SCHEME.equals(uri.getScheme())) {
            return null;
        }
        String sitemap = uri.getPath();
        String item = TextUtils.isEmpty(uri.getQueryParameter(DEPRECATED_QUERY_PARAMETER_ITEM_NAME))
                ? uri.getQueryParameter(QUERY_PARAMETER_ITEM_NAME)
                : uri.getQueryParameter(DEPRECATED_QUERY_PARAMETER_ITEM_NAME);
        String label = uri.getQueryParameter(QUERY_PARAMETER_ITEM_LABEL);
        String state = TextUtils.isEmpty(uri.getQueryParameter(DEPRECATED_QUERY_PARAMETER_STATE))
                ? uri.getQueryParameter(QUERY_PARAMETER_STATE)
                : uri.getQueryParameter(DEPRECATED_QUERY_PARAMETER_STATE);
        String mappedState = uri.getQueryParameter(QUERY_PARAMETER_MAPPED_STATE);

        return new AutoValue_NfcTag.Builder()
                .sitemap(!TextUtils.isEmpty(sitemap) && TextUtils.isEmpty(item) ? sitemap : null)
                .item(item)
                .label(label)
                .state(state)
                .mappedState(mappedState)
                .build();
    }
}

File: mobile/src/main/java/org/openhab/habdroid/model/NfcTag.kt
/*
 * Copyright (c) 2010-2020 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.model

import android.net.Uri

data class NfcTag(
    val sitemap: String?,
    val item: String?,
    val label: String?,
    val state: String?,
    val mappedState: String?
) {
    companion object {
        const val SCHEME = "openhab"
        const val QUERY_PARAMETER_ITEM_NAME = "i"
        const val DEPRECATED_QUERY_PARAMETER_ITEM_NAME = "item"
        const val QUERY_PARAMETER_STATE = "s"
        const val DEPRECATED_QUERY_PARAMETER_STATE = "command"
        const val QUERY_PARAMETER_MAPPED_STATE = "m"
        const val QUERY_PARAMETER_ITEM_LABEL = "l"
    }
}

fun Uri.toTagData(): NfcTag? {
    if (scheme != NfcTag.SCHEME) {
        return null
    }
    val item = if (NfcTag.DEPRECATED_QUERY_PARAMETER_ITEM_NAME in queryParameterNames)
        getQueryParameter(NfcTag.DEPRECATED_QUERY_PARAMETER_ITEM_NAME)
    else
        getQueryParameter(NfcTag.QUERY_PARAMETER_ITEM_NAME)

    val state = if (NfcTag.DEPRECATED_QUERY_PARAMETER_STATE in queryParameterNames)
        getQueryParameter(NfcTag.DEPRECATED_QUERY_PARAMETER_STATE)
    else
        getQueryParameter(NfcTag.QUERY_PARAMETER_STATE)

    val label = getQueryParameter(NfcTag.QUERY_PARAMETER_ITEM_LABEL)
    val mappedState = getQueryParameter(NfcTag.QUERY_PARAMETER_MAPPED_STATE)
    val sitemapPath = path
    val sitemap = if (sitemapPath?.isNotEmpty() == true) sitemapPath else null

    return NfcTag(sitemap, item, label, state, mappedState)
}
